src/Serilog/Sinks/PeriodicBatching/PeriodicBatchingSink.cs

```diff
@@ -216,11 +213,8 @@ void OnTick()
                     while (_queue.TryDequeue(out evt)) { }
                 }
 
-                lock (_stateLock)
-                {
-                    if (!_unloading)
-                        SetTimer(_status.NextInterval);
-                }
+                if (_state == Started)
```

Comment:
I agree it is tricky place here, I spent whole day refreshing ECMA memory model and diffs between lock vs volatile vs Volatile.Read vs Thread.VolatileRead.

(I wrote a lot trying to back up this line, but then I understand something and remove it)

So, kind of "racing" can happen, even when checking state under lock. I mean other thread can be 1ms late setting unloading and then we have timer scheduled for one more tick.

And give that this line is situated in OnTick method, which is executed in another (timer's) thread and does'nt concern us in terms of contention, here we can use Thread.VolatileRead as strongest fence to read latest value.

This also gives possibility to remove volatile keyword from _state and avoid _possible_ compiler warning mentioned in comment above.